distage/distage-roles-api/src/main/scala/com/github/pshirshov/izumi/distage/roles/roles/ResourceCollection.scala

```diff
@@ -0,0 +1,22 @@
+package com.github.pshirshov.izumi.distage.roles.roles
+
+import java.util.concurrent.ExecutorService
+
+import com.github.pshirshov.izumi.distage.model.Locator
+import com.github.pshirshov.izumi.distage.model.plan.ExecutableOp
+
+trait ResourceCollection {
+  def startMemoizedComponents(): Unit
+
+  def getCloseables: Set[AutoCloseable]
+
+  def getComponents: Set[RoleComponent]
+
+  def getExecutors: Set[ExecutorService]
+
+  def processContext(context: Locator): Unit
+
+  def transformPlanElement(op: ExecutableOp): ExecutableOp
+
+  def close(closeable: AutoCloseable): Unit
```

Comment:
This method is never used, and also `getX` methods expose internals in an extreme way making it *even* harder to refactor in the future than currently.

I propose to change the interface to just

```scala
trait ResourceCollection {
  def canStart(resource: Any): Boolean

  def registerStarted(resource: Any): Unit

  def canClose(resource: Any): Boolean

  def processContext(context: Locator): Unit

  def transformPlanElement(op: ExecutableOp): ExecutableOp
}
```

This should drastically simplify logic - `ResourceCollection` wouldn't need to duplicate any of the lifecycle functions already done by `RoleStarter`, `RoleStarter` would just need to query `ResourceCollection` before starting/stopping any AutoCloseables/RoleComponents/ExecutorServices and call `registerStarted` after starts - where `registerStarted` will do nothing for unmemoized instances, but start returning `canStart=false` for memoized instances.